subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -1065,7 +1086,23 @@ static void sm_do_network_error(void)
 	}
 #endif
 
-	set_sm_state(ENGINE_DO_REGISTRATION);
+	if (!client.last_update || (k_uptime_get() - client.last_update) / 1000 > client.lifetime) {
+		/* do full registration as there is no active registration or lifetime exceeded */
+		set_sm_state(ENGINE_DO_REGISTRATION);
+		return;
+	}
+
+	err = lwm2m_socket_start(client.ctx);
+	if (err) {
+		LOG_ERR("Failed to start socket %d", err);
+		/* keep this state until lifetime/retry count exceeds. Renew sm state
+		 * to set retry_delay etc ... */
+		set_sm_state(ENGINE_NETWORK_ERROR);
+		return;
+	}
+
+	engine_trigger_update(false);
+	set_sm_state(ENGINE_REGISTRATION_DONE);
```

Comment:
This state change worries me.
`NETWORK_ERROR ---(invalid state change)---> REGISTRATION_DONE --> UPDATE_SENT`

I see that by doing this, you are triggering the functionality in `sm_registration_done()`. But this is now hard to spot, and not obvious at all for developer looking at this code. If we need calls from `sm_registration_done()` maybe it needs to split out from there, so we can directly call the function and stay correctly in `NETWORK_ERROR` state. Or maybe we should consider going into `RX_OFF` or something like that.